mesmerize_core/batch_utils.py

```diff
@@ -245,3 +247,68 @@ def get_full_raw_data_path(path: Union[Path, str]) -> Path:
         return PARENT_DATA_PATH.joinpath(path)
 
     return path
+
+
+class OverwriteError(IndexError):
+    """
+    Error thrown when trying to write to an existing batch file, but there is a risk
+    of overwriting existing data.
+    Note this is a subclass of IndexError to avoid a breaking change, because
+    IndexError was previously thrown from df.caiman.save_to_disk.
+    """
+    pass
+
+
+class BatchLock:
+    """Locks a batch file for safe writing, returning the dataframe in the target"""
+    TIMEOUT = 30  # must be consistent or else nested re-entrant locks break
+    def __init__(self, batch_path: Union[Path, str]):
+        self.lock = SoftFileLock(str(batch_path) + ".lock", is_singleton=True, timeout=self.TIMEOUT)
+        self.batch_path = batch_path
+    
+    def __enter__(self) -> pd.DataFrame:
+        self.lock.__enter__()   # may throw Timeout
+        return load_batch(self.batch_path)
+    
+    def __exit__(self, exc_type, exc_value, traceback):
+        self.lock.__exit__(exc_type, exc_value, traceback)
+
+
+def open_batch_for_safe_writing(batch_path: Union[Path, str]) -> BatchLock:
+    """Just a more self-documenting constructor"""
+    return BatchLock(batch_path)
+
+
+def save_results_safely(batch_path: Union[Path, str], uuid, results: dict, runtime: float):
+    """
+    Try to load the given batch and save results to the given item
+    Uses a file lock to ensure that no other process is writing to the same batch using this function,
+    which gives up after lock_timeout seconds (set to -1 to never give up)
+    """
+    try:
+        with open_batch_for_safe_writing(batch_path) as df:
+            # Add dictionary to output column of series
+            df.loc[df["uuid"] == uuid, "outputs"] = [results]
+            # Add ran timestamp to ran_time column of series
+            df.loc[df["uuid"] == uuid, "ran_time"] = datetime.now().isoformat(timespec="seconds", sep="T")
+            df.loc[df["uuid"] == uuid, "algo_duration"] = str(runtime) + " sec"
+           
+            # save dataframe to disk
+            df.caiman.save_to_disk()
+
+    except BaseException as e:
```

Comment:
```suggestion
except (Timeout, PreventOverwriteError) as e:
```

`BaseException` should never be used AFAIK

Reply:
Probably `Exception` is better than `BaseException` as we wouldn't want keyboard interrupts to be treated in the same way. But my intention was to make sure a runtime error doesn't get lost if there is a failure to write to the batch file for *any* reason.

This is less important if we do always log errors to some file within the item directory, so even if saving fails, that can be inspected.

Reply:
Ah I see, then yes use broad Exception instead of base